File: src/lab4/java/SelectionSort.java
package lab4.java;

public class SelectionSort implements SortAlgorithm {

    private int[] elems;
    private int actualIter = 0;
    private int actualPos = actualIter + 1;
    private int minimumPos = actualIter;

    public SelectionSort(int[] elems) {
        this.elems = elems;
    }

    @Override
    public void step(Swap s) {
        if (elems[actualPos] < elems[minimumPos]) {
            minimumPos = actualPos;
        }

        if (actualPos == elems.length - 1) {
            if (actualIter != minimumPos) {
                s.swap(actualIter, minimumPos);
                int temp = elems[minimumPos];
                elems[minimumPos] = elems[actualIter];
                elems[actualIter] = temp;
            }

            actualIter++;
            actualPos = actualIter + 1;
            minimumPos = actualIter;
        } else {
            actualPos++;
        }

    }

    @Override
    public boolean isSorted() {
        return actualIter == elems.length - 1;
    }

}

File: src/lab4/SelectionSort.kt
package lab4

class SelectionSort(override val elems: Array<Int>) : SortAlgorithm {

    private var actualIter = 0
    private var actualPos = actualIter + 1
    private var minimumPos = actualIter

    override fun step(doRepaint: () -> Unit) {

        if (elems[actualPos] < elems[minimumPos]) {
            minimumPos = actualPos
        }

        if (actualPos == elems.size - 1) {
            if (actualIter != minimumPos) {
                doRepaint()
                val temp = elems[minimumPos]
                elems[minimumPos] = elems[actualIter]
                elems[actualIter] = temp
            }

            actualIter++
            actualPos = actualIter + 1
            minimumPos = actualIter
        } else {
            actualPos++
        }

    }

    override fun isSorted() = actualIter == elems.size - 1

}
